crates/evm-prover/src/prover/prover.rs

```diff
@@ -66,15 +87,214 @@ pub trait ProgramProver {
+        })
+    }
+}
+
 /// A prover for generating SP1 proofs for EVM block execution and data availability in Celestia.
 ///
 /// This struct is responsible for preparing the standard input (`SP1Stdin`)
 /// for a zkVM program that takes a blob inclusion proof, data root proof, Celestia Header and
 /// EVM state transition function.
 pub struct BlockExecProver {
-    config: Config,
-    prover: EnvProver,
-    // extend with custom state, e.g. celestia rpc, evm rpc, etc...
+    pub app: AppContext,
+    pub config: ProverConfig,
+    pub prover: EnvProver,
+}
+
+impl BlockExecProver {
+    /// Creates a new instance of [`BlockExecProver`] for the provided [`AppContext`] using default configuration
+    /// and prover environment settings.
+    pub fn new(app: AppContext) -> Self {
+        let config = BlockExecProver::default_config();
+        let prover = ProverClient::from_env();
+
+        Self { app, config, prover }
+    }
+
+    /// Returns the default prover configuration for the block execution program.
+    pub fn default_config() -> ProverConfig {
+        ProverConfig {
+            elf: EVM_EXEC_ELF,
+            proof_mode: SP1ProofMode::Compressed,
+        }
+    }
+
+    /// Generates the serialized state transition function (STF) input for a given EVM block number.
+    async fn generate_stf(&self, block_number: u64) -> Result<Vec<u8>> {
+        let host_executor = EthHostExecutor::eth(self.app.chain_spec.clone(), None);
+        let provider = ProviderBuilder::new().on_http(self.app.evm_rpc.parse()?);
+        let rpc_db = RpcDb::new(provider.clone(), block_number - 1);
+
+        let client_input = host_executor
+            .execute(block_number, &rpc_db, &provider, self.app.genesis.clone(), None, false)
+            .await?;
+
+        Ok(bincode::serialize(&client_input)?)
+    }
+
+    /// Queries the inclusion height for the given EVM block number from the sequencer rpc.
+    async fn inclusion_height(&self, block_number: u64) -> Result<u64> {
+        let mut client = StoreServiceClient::connect(self.app.sequencer_rpc.clone()).await?;
+        let req = GetMetadataRequest {
+            key: format!("rhb/{}/d", block_number),
+        };
+
+        let resp = client.get_metadata(req).await?;
+        let height = u64::from_le_bytes(resp.into_inner().value[..8].try_into()?);
+
+        Ok(height)
+    }
+
+    /// Queries the extended header from the Celestia data availability rpc.
+    async fn extended_header(&self, height: u64) -> Result<ExtendedHeader> {
+        let client = Client::new(&self.app.celestia_rpc, None).await?;
+        let header = client.header_get_by_height(height).await?;
+
+        Ok(header)
+    }
+
+    /// Queries all blobs from Celestia for the given inclusion height and filters them to find the blob for the
+    /// corresponding EVM block number.
+    async fn blob_for_height(&self, block_number: u64, inclusion_height: u64) -> Result<Blob> {
```

Comment:
This function needs some follow up work, as I learned today that rollkit does not publish `SignedData` for block without any txs.

Reply:
I opened https://github.com/celestiaorg/celestia-zkevm-hl-testnet/issues/79 and can address in a follow up so maybe we can just merge this PR